src/js/components/FileInput/FileInput.js

```diff
@@ -0,0 +1,171 @@
+import React from 'react';
+import { compose } from 'recompose';
+import styled, { withTheme } from 'styled-components';
+import { FormClose } from 'grommet-icons';
+
+import { defaultProps } from '../../default-props';
+
+import { withFocus, withForwardRef } from '../hocs';
+import { focusStyle } from '../../utils';
+
+import { Anchor } from '../Anchor';
+import { Box } from '../Box';
+import { Button } from '../Button';
+import { Keyboard } from '../Keyboard';
+import { Stack } from '../Stack';
+import { Text } from '../Text';
+
+import { StyledFileInput } from './StyledFileInput';
+
+const ControlBox = styled(Box)`
+  cursor: pointer;
+  ${props => props.focus && focusStyle}
+`;
+
+const FileInput = ({
+  a11yTitle,
+  focus,
+  forwardRef,
+  messages,
+  theme,
+  onChange,
+  ...rest
+}) => {
+  const [hover, setHover] = React.useState();
+  const [files, setFiles] = React.useState([]);
+  const internalRef = React.useRef();
+
+  return (
+    <Keyboard
+      onSpace={event => {
+        if (event.currentTarget === event.target) {
+          (forwardRef || internalRef).current.click();
+        }
+      }}
+      onEnter={event => {
+        if (event.currentTarget === event.target) {
+          (forwardRef || internalRef).current.click();
+        }
+      }}
+    >
+      <Stack
+        a11yTitle={a11yTitle}
+        tabIndex={0}
+        guidingChild="last"
+        interactiveChild={files.length > 0 ? 'last' : 'first'}
+        onClick={() => (forwardRef || internalRef).current.click()}
+        onMouseOver={() => setHover(true)}
+        onMouseOut={() => setHover(false)}
+        onFocus={() => setHover(false)}
+        onBlur={() => setHover(false)}
+      >
+        <StyledFileInput
+          ref={forwardRef || internalRef}
+          tabIndex={-1}
+          {...rest}
+          type="file"
+          onChange={event => {
+            event.persist();
+            const fileList = event.target.files;
+            const nextFiles = [...files];
+            for (let i = 0; i < fileList.length; i += 1) {
+              nextFiles.push(fileList[i]);
+            }
+            setFiles(nextFiles);
+            if (onChange) {
+              onChange(event);
+            }
+          }}
+        />
+
+        <ControlBox
+          border={{
+            color: hover ? 'brand' : undefined,
+            side: 'all',
+            size: 'medium',
+            style: 'dashed',
+          }}
+          align={files.length ? 'stretch' : 'center'}
+          justify="center"
+        >
+          {files.length > 10 && (
```

Comment:
why do we use `10` as a restriction? 
I see it is being used in line #109 as well, I think we should extract it to a variable with a name that implies the restriction reasoning.

Reply:
good idea, added `aggregateThreshold` for this.